torch/distributed/_tensor/dispatch.py

```diff
@@ -149,14 +151,72 @@ def operator_dispatch(
-        # a consistent way to handle that, and also need to figure
-        # out if we should communicate the result to non-participating
-        # ranks (i.e. a.sum() -> scalar, maybe we should set to 0)
-        local_results = torch.tensor([])
+        # For a non-participating device, we do:
+        # 1. if the return type is scalar, all gather the local result
+        # from participating devices, and reduce on the list of results.
+        # For bool type, we by default use AND to reduce;
+        # We can extend this logic for specific ops if necessary.
+        # 2. if the return type is Tensor, return empty tensor with
+        # correct dtype.
+        # 3. if the return type is List[Tensor], return a list of empty
+        # tensors with correct dtype.
+        spec = output_sharding.output_spec
+        ret_list = op_schema.func_schema.returns
+        if len(ret_list) != 1:
+            # returns list should only have one Argument
+            raise NotImplementedError(
+                f"function schema {str(op_schema.func_schema)} has"
+                f" return type that we currently don't support."
+            )
+
+        if spec is None:
+            # return a scalar value
+            # collect local results from participating ranks
+            obj_list = [None for _ in range(dist.get_world_size())]
+            dist.all_gather_object(obj_list, None)
+            obj_list = list(filter(lambda x: x is not None, obj_list))
+            # perform reduce on the collection with AND op
+            ret_type = str(ret_list[0].type)
+            if ret_type == "bool":
+                import operator
+                local_results = functools.reduce(operator.and_, obj_list, True)
+            else:
+                raise NotImplementedError(
+                    f"default value for type {ret_type} on non-participating"
+                    f" device is not implemented."
+                )
+        else:
+            def default_tensor(spec: DTensorSpec):
+                if spec.tensor_meta is not None:
+                    shape = spec.tensor_meta.shape
+                    dtype = spec.tensor_meta.dtype
+                    if len(shape) == 0:
+                        # scalar tensor
+                        return torch.zeros((), dtype=dtype)
+                    else:
+                        # non-scalar tensor
+                        return torch.tensor([], dtype=dtype)
+                else:
+                    raise RuntimeError(
+                        f"{spec} has no tensor metadata."
+                    )
+
+            if (isinstance(spec, DTensorSpec)):
+                # return a Tensor value
+                local_results = default_tensor(spec)
+            elif (isinstance(spec, Sequence)):
+                # return a List[Tensor] value
+                local_results = []
+                for s in spec:
+                    if s is not None:
+                        local_results.append(default_tensor(s))
```

Comment:
nit: `local_results = [default_tensor(s) if s is not None else None for s in spec]`

Reply:
My confusion right now is, can we safely assume the return List has only Tensor objects? E.g. lstm_mps_backward (
https://github.com/pytorch/pytorch/blob/cc699c56dc14af5b652a49e3044c0fe7c8ca70ee/aten/src/ATen/native/native_functions.yaml#L7222-L7225). I checked our op registration and all multiple returns ops only return multiple Tensors.

Reply:
we don't need to worry about one corner case as we can always deal with that later when we actually add that op

Reply:
agree. then I'll just add
```
if None in local_results:
    raise NotImplementedError(...)
```